crates/astria-sequencer/src/mempool/mod.rs

```diff
@@ -267,34 +286,104 @@ impl Mempool {
-
-        // clean accounts of stale and expired transactions
-        let mut removed_txs = pending.clean_accounts(&current_account_nonce_getter).await;
-        removed_txs.append(&mut parked.clean_accounts(&current_account_nonce_getter).await);
-
-        // run promotion logic in case transactions not in this mempool advanced account state
-        let to_promote = parked.find_promotables(&current_account_nonce_getter).await;
-        // Release the lock asap.
-        drop(parked);
-        for (ttx, current_account_nonce) in to_promote {
-            if let Err(error) = pending.add(ttx, current_account_nonce) {
-                error!(
-                    current_account_nonce,
-                    "failed to promote transaction during maintenance: {error:#}"
-                );
+        let mut removed_txs = Vec::<([u8; 32], RemovalReason)>::new();
+
+        // To clean we need to:
+        // 1.) remove stale and expired transactions from both pending and parked
+        // 2.) check if we have transactions in pending which need to be demoted due
+        //     to balance decreases
+        // 3.) if there were no demotions, check if parked has transactions we can
+        //     promote
+
+        // TODO reprice transactions also if FeeAssetChange or FeeChange Actions were ran :0
+
+        let addresses: HashSet<[u8; 20]> = pending
+            .addresses()
+            .into_iter()
+            .chain(parked.addresses())
+            .collect();
+
+        for address in addresses {
+            // get current account state
+            let current_nonce = match current_account_nonce_getter(address).await {
+                Ok(res) => res,
+                Err(error) => {
+                    error!(
+                        address = %telemetry::display::base64(&address),
+                        "failed to fetch account nonce when cleaning accounts: {error:#}"
+                    );
+                    continue;
+                }
+            };
+            let current_balances = match current_account_balances_getter(address).await {
+                Ok(res) => res,
+                Err(error) => {
+                    error!(
+                        address = %telemetry::display::base64(&address),
+                        "failed to fetch account balances when cleaning accounts: {error:#}"
+                    );
+                    continue;
+                }
+            };
+
+            // clean pending and parked of stale and expired
+            removed_txs.extend(pending.clean_account_stale_expired(address, current_nonce));
+            removed_txs.extend(parked.clean_account_stale_expired(address, current_nonce));
+
+            // get transactions to demote from pending
+            let demotion_txs = pending.find_demotables(address, &current_balances);
+
+            if demotion_txs.is_empty() {
+                // nothing to demote, check for transactions to promote
```

Comment:
why do we only promote txs if there's nothing to demote? shouldn't both occur either way?

Reply:
If we demoted something from an account's pending queue that means there shouldn't be anything to promote from the account's parked queue. The pending queue's enforcement of sequential nonces is the main reason for this.

For example, if we have nonces `[0,1,2]` in pending and nonces `[3,4]` in parked , and `[2]` gets demoted from pending because of lack of account funds, then parked would be `[2,3,4]`.  During the promotion loop `2` would be ineligible for promotion due to lack of account funds, also making `3,4` ineligible because of the sequential nonce requirement.